### PR TITLE
Add version check command

### PR DESCRIPTION
**Pull request by 0x48piraj**
| | |
|:---|:---|
|**_Related Issue(s)_**| https://github.com/nasa/fprime/issues/2407  |
|**_Has Unit Tests (y/n)_**| n |
|**_Documentation Included (y/n)_**| y |

---
## Change Description

Adds a version check command for

- Python version
- Pip version
- CMake version and
- Toolchain version info

## Rationale

Help debug things

## Future Work

Unit tests preferably.

### CODE REVIEW

File: src/fprime/util/commands.py [hunk 1]
```diff
@@ -182,3 +184,53 @@ def run_code_format(
...
+    except ModuleNotFoundError: # Python >=3.6
+        print("[WARNING] Cannot import 'Pip'.")
+
+    try:
+        import pkg_resources
+    except ModuleNotFoundError: # Python >=3.6
+        print("[WARNING] Cannot import 'pkg_resources'. Will not check tool versions.")
+        return
+
+    tools = [
```
Comment (thomas-bc) [inline src/fprime/util/commands.py]: This list is the same as here: https://github.com/0x48piraj/fprime-tools/blob/863bde64a8965b881d1fdddd4729437e10d9d4b6/src/fprime/util/build_helper.py#L71-L85

Could we define this at the module level in build_helper.py, and import that variable here, instead of re-defining it?

File: src/fprime/util/cli.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 from fprime.fbuild.target import Target
 from fprime.fpp.cli import add_fpp_parsers
 from fprime.util.build_helper import load_build
-from fprime.util.commands import run_code_format, run_hash_to_file, run_info, run_new
+from fprime.util.commands import run_code_format, run_hash_to_file, run_info, run_new, run_sysinfo
 from fprime.util.help_text import HelpText
 from fprime.fpp.visualize import add_fpp_viz_parsers
 from fprime.fpp.impl import add_fpp_impl_parsers
```
Comment (thomas-bc) [inline src/fprime/util/cli.py:28]: In this file, you should be able to add the new command to the [skip_build_loading()](https://github.com/0x48piraj/fprime-tools/blob/863bde64a8965b881d1fdddd4729437e10d9d4b6/src/fprime/util/cli.py#L53-L59) function 

This will skip trying to load the Build object (see [here](https://github.com/0x48piraj/fprime-tools/blob/863bde64a8965b881d1fdddd4729437e10d9d4b6/src/fprime/util/cli.py#L30-L34)), and we'll be able to run this command form anywhere, no need to be in an F´ project

The following should do the trick
```
    if parsed.command == "version-check":
        return True
```

File: src/fprime/util/commands.py [hunk 3]
```diff
@@ -182,3 +184,53 @@ def run_code_format(
...
+        "fprime-fpp-from-xml",
+        "fprime-fpp-format",
+        "fprime-fpp-filenames",
+        "fprime-fpp-depend",
+        "fprime-fpp-check",
+    ]
+    for tool in tools:
+            try:
+                version = pkg_resources.get_distribution(tool).version
+                print(f"{tool} version: {version}")
```
Comment (thomas-bc) [inline src/fprime/util/commands.py]: Can we indent and reformat so the output looks something like this:
```
Python version: 310.12
CMake version: 3.24.2
Pip version: 23.0.1
Pip packages:
    fprime-tools==3.4.2
    fprime-gds==3.3.4
    fprime-fpp-to-xml==2.0.2
    fprime-fpp-to-json==2.0.2
    fprime-fpp-to-cpp==2.0.2
    fprime-fpp-syntax==2.0.2
    fprime-fpp-locate-uses==2.0.2
    fprime-fpp-locate-defs==2.0.2
    fprime-fpp-from-xml==2.0.2
    fprime-fpp-format==2.0.2
    fprime-fpp-filenames==2.0.2
    fprime-fpp-depend==2.0.2
    fprime-fpp-check==2.0.2
```

**Comment by 0x48piraj** [comment]
I also added a skeleton unit test for commands (as asked above in the PR), it's pretty empty as of now but I would be happy to write tests for `special-parser` commands.

File: src/fprime/util/cli.py [hunk 4]
```diff
@@ -211,6 +220,7 @@ def add_special_parsers(
     return {
         "hash-to-file": run_hash_to_file,
         "info": run_info,
+        "sysinfo": run_sysinfo,
```
Comment (thomas-bc) [inline src/fprime/util/cli.py]: Can we rename the CLI target to be `version-check` instead of `sysinfo` ?

**Review by thomas-bc — CHANGES_REQUESTED** [review]
This is awesome, thank you very much!

Looks good to me - just requesting a couple of very minor changes on formatting etc...

Also make sure to take a look at the spellings that were flagged by CI and to add them in https://github.com/0x48piraj/fprime-tools/blob/fd5add0f8781825ea90bfccdcd8d3fb1872d36b2/.github/actions/spelling/expect.txt if they are false positives.

**Comment by thomas-bc** [comment]
Looks like you also need to run the formatter -
```
pip install black
black ./path/to/file_to_format.py
```

I figured you might want to take this across the finish line yourself, but let me know if you want me to help make some changes.

File: src/fprime/util/commands.py (hunk 1 above)
Comment (0x48piraj) [reply src/fprime/util/commands.py]: How about we take the variable from **build_helper.py** and put it over **versioning.py** inside a function called `get_toolchain` which returns the list and use it over build-helper and commands?

Comment (thomas-bc) [reply src/fprime/util/commands.py]: Yes even better, I wouldn't put it in a function though, just FPRIME_PIP_PACKAGES global variable or something

**Comment by 0x48piraj** [comment]
All the requested changes were made including the unused import one and it works as intended in local settings.

**Comment by thomas-bc** [comment]
Thanks! Note to myself: add OS and processor architecture information as well

File: test/fprime/util/commands_unit_test.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+(test) fprime-utils commands:
+
+Tests the F prime util commands.
+@author 0x48piraj
+"""
+
+import unittest
+
+
+class CommandsTestCases(unittest.TestCase):
```
Comment (LeStarch) [inline test/fprime/util/commands_unit_test.py:11]: What are there a bunch of empty tests here?

File: src/fprime/util/commands.py [hunk 6]
```diff
@@ -182,3 +184,47 @@ def run_code_format(
...
     return clang_formatter.execute(build, parsed.path, ({}, parsed.pass_through))
+
+
+def run_version_check(
+    base: Build, parsed: argparse.Namespace, _: Dict[str, str], __: Dict[str, str], ___
+):
+    """Print out versions to help debugging"""
+
+    print(
+        f"Python version: {sys.version_info.major}{sys.version_info.minor}.{sys.version_info.micro}"
```
Comment (LeStarch) [inline src/fprime/util/commands.py]: Add a "." between major and minor version,

File: src/fprime/util/commands.py [hunk 7]
```diff
@@ -182,3 +184,47 @@ def run_code_format(
     for filename in parsed.files:
         clang_formatter.stage_file(Path(filename))
     return clang_formatter.execute(build, parsed.path, ({}, parsed.pass_through))
+
+
+def run_version_check(
+    base: Build, parsed: argparse.Namespace, _: Dict[str, str], __: Dict[str, str], ___
+):
+    """Print out versions to help debugging"""
+
```
Comment (LeStarch) [inline src/fprime/util/commands.py:193]: Would you be up for adding os and chip architecture here (in the python `platform` package?

**Review by LeStarch — DISMISSED** [review]
A few minor nits, but this is great work!

File: src/fprime/util/commands.py (hunk 6 above)
Comment (0x48piraj) [reply src/fprime/util/commands.py]: Ah! I'm so sorry!

File: test/fprime/util/commands_unit_test.py (hunk 5 above)
Comment (0x48piraj) [reply test/fprime/util/commands_unit_test.py:11]: As commented above, it's a skeleton structure and I was thinking of writing tests separately in another pull, just making a placeholder class.

File: src/fprime/util/commands.py (hunk 7 above)
Comment (0x48piraj) [reply src/fprime/util/commands.py:193]: Sure!

**Comment by 0x48piraj** [comment]
Rather than using sys, I utilized the `platform` library for retrieving all system info.

**Comment by 0x48piraj** [comment]
Any updates on this @thomas-bc?

File: test/fprime/util/commands_unit_test.py (hunk 5 above)
Comment (thomas-bc) [reply test/fprime/util/commands_unit_test.py:11]: @0x48piraj if you start looking into adding test cases, we usually use `pytest` over `unittest`

**Review by thomas-bc — APPROVED** [review]

**Review by LeStarch — APPROVED** [review]
Looks good!

**Comment by thomas-bc** [comment]
Apologies for the delay, I was out for a bit. LGTM! Thank you very much for this, we want to add this command to our bug_report.md issue template over on https://github.com/nasa/fprime. Feel free to make the PR yourself there if you want to! e.g. add a section/checkmark

> [] Please paste the output of fprime-util version-check:
